### ISSUE TITLE
Notebook 07: ReLU and why neural networks need nonlinearity

### ISSUE THREAD

**Issue by majorgilles** [majorgilles/pure-python-mlp-from-scratch#17]
## Learning goal

Explain why a hidden layer needs a nonlinearity and make ReLU’s forward pass and derivative easy to remember.

## What to build

Create a notebook that introduces ReLU:

```text
relu(x) = max(0, x)
```

and its derivative:

```text
1 when x > 0, otherwise 0
```

## Suggested flow

1. Show that stacking linear layers without nonlinearity is still linear.
2. Introduce ReLU as a simple “gate.”
3. Evaluate ReLU for negative, zero, and positive inputs.
4. Explain the derivative as whether the gate is open.
5. Use tiny examples to show how gradients stop when ReLU output is zero.

## Acceptance criteria

- [ ] Notebook explains nonlinearity in plain language.
- [ ] Notebook includes a small ReLU table for inputs like `-2`, `0`, `3`.
- [ ] Notebook derives or explains the ReLU derivative.
- [ ] Notebook shows how ReLU changes backward flow through one hidden value.
- [ ] Notebook states the convention used at exactly `x = 0` for this course.
- [ ] Notebook includes a checkpoint where the learner manually computes ReLU forward/backward values.
- [ ] All code uses only the Python standard library.

## Resources to follow

- PyTorch ReLU docs for later naming alignment: https://docs.pytorch.org/docs/stable/generated/torch.nn.ReLU.html
- Deep Learning Book MLP chapter: https://www.deeplearningbook.org/contents/mlp.html
- Karpathy micrograd examples for ReLU-style operations: https://github.com/karpathy/micrograd

## Blocked by

- Blocked by #16
